### DISCUSSION TITLE
[@types/react] KeyboardEvent

### DISCUSSION THREAD

**Discussion by SWBennett06** [DefinitelyTyped/DefinitelyTyped#62608] [Issues with a @types package] [answered]
`KeyboardEvent<HTMLInputElement>` is missing properties I would expect from the HTMLInputElement type (also noticed with `HTMLTextAreaElement`):

```ts
function handleKeyDown(e: KeyboardEvent<HTMLInputElement>) {
  const { value } = e.target; // <--- throws "Property 'value' does not exist on type 'EventTarget'.ts(2339)"
}
```

The type on hover of `target` is:

```
(property) React.BaseSyntheticEvent<globalThis.KeyboardEvent, EventTarget & HTMLInputElement, EventTarget>.target: EventTarget
```

The same destructuring assignment presents no error for `ChangeEvent<HTMLInputElement>`, where the type shows as:

```
(property) React.ChangeEvent<HTMLInputElement>.target: EventTarget & HTMLInputElement
```

It looks like `KeyboardEvent<T>` is not correctly pulling in the types on the given type parameter.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
Events bubble in the DOM. So the `target` might be some nested HTML element i.e. TypeScript can't know the possible types at compile time other than `EventTarget`.

You might be interested in `currentTarget` instead whose type we do know at compile time since it's the same type the event handlers is attached to.

**Comment by SWBennett06** [comment]
Ah, so I assume `target` is behaving as I expected for React's `ChangeEvent` because the only possible emitters are `input`, `textarea`, or `select`, which can be guaranteed to have the given properties? 

On the other hand, the `input` element is a void element which cannot have children from which an event might bubble, so should have a similar guarantee.

Regardless, I can move forward with `currentTarget`. Thanks!
